Make autoSort toggleable

JorisGoosen opened the pull request:
Implements https://github.com/jasp-stats/INTERNAL-jasp/issues/2555 and makes the alignment of the labeleditor better as well

JorisGoosen committed:
remove useless header includes

JorisGoosen committed:
Add autosort by value to jasp

Implements https://github.com/jasp-stats/INTERNAL-jasp/issues/2555

Also tweaks the display of variableswindow to be a bit better aligned
breaks loading of 0.19 files made before this commit

JorisGoosen committed:
autoSort should be disablable

also makes Ctrl+z etc alwys enabled, because you can always make all sorts of changes we now have undo tracking for

JorisGoosen committed:
add some better behaviour for editing (original)value of a label

if value and label are the same then on a value edit:
- if numbers and new number or text as value, doesnt matter replace them both.
- if not numbers and a nonnumber is entered then again it is assumed that also the label needs to change
- if not numbers, then on a new number it only changes the value, not the label. this allows for recoding

JorisGoosen committed:
fix cpp.in too

JorisGoosen committed:
add property to columnModel: firstNonNumericRow

JorisGoosen committed:
try to make undo setData much more secure

JorisGoosen committed:
fix lines

JorisGoosen committed:
fix reverse values

JorisGoosen committed:
better buttons and tooltips

JorisGoosen committed:
fix label undos!

JorisGoosen committed:
make buttons in labelseditor be enabled and/or visible at the perfect times

JorisGoosen committed:
better undo for setData finished

also make sure move-label-up-button is disabled when no autoSort and in the top

JorisGoosen committed:
count numerics for https://github.com/jasp-stats/INTERNAL-jasp/issues/2552

boutinb commented:
I have still some strange things... Open debug.csv:

- Open contBinom Label Editor 
- Unset 'Automatically order labels by their numeric values'
- Undo: The values get a decimal format (0.0 and 1.0)

boutinb commented:
Also:

- Open contBinom Label Editor
- Unset 'Automatically order labels by their numeric values'
- Select the first row (value 0)
- Click on the 'Move selected labels down'
- The row goes down, but the buttons to move the labels up or down are disabled.

JorisGoosen committed:
fix label deserialize and renamed a field to intsId to be less misleading

JorisGoosen committed:
fix move-label-buttons after a move

JorisGoosen committed:
stop edit-item from randomly jumping around when you click outside of it to stop editing

before this change, when editing (for instance) the third element of contBinom and you click another cell, it randomly jumps to a wrong one entirely.
this is caused by the reusing of items due to turning caching back on.
But this code fixes the actual problem by not moving the item with the mousearea around right in the middle of the click.

JorisGoosen commented:
those label problems have been fixed (there was an issue where label deserialisation went wrong because the original-value-json was being forced to string instead of it just staying as it should be (a json:value!).

Also fixed a bug caused by turning caching of items in the datasetview, where a random edit item would be opened when trying to change editposition by clicking.

boutinb commented on Desktop/components/JASP/Widgets/JASPDataView.qml:
```diff
@@ -16,6 +16,7 @@ FocusScope
 				property alias mouseArea:				wheelCatcher
 				property bool  doubleClickWorkaround:	true
 				property alias isMainDataViewer:		theView.mainData
+	readonly	property alias editCoordinates:			theView.editCoordinates
```
I don't see where editCoordinates is used. Can be removed?

JorisGoosen replied:
Could be removed, I just added it because it might be useful at some point and its not too much extra code anyway.
I first thought the variable-editor used this too...

JorisGoosen replied:
Im not removing it, it could turn out to be useful

boutinb commented on Desktop/data/undostack.h:
```diff
@@ -58,35 +58,44 @@ class SetWorkspacePropertyCommand: public UndoModelCommand
 };
 
-class SetLabelOriginalValueCommand: public UndoModelCommand
+class SetLabelOriginalValueCommand: public UndoModelCommandLabelChange
```
I think you forgot to also undo value changes in the data. The undo method must be also overriden.
If you change the value of a label, and undo it, it does not undo the values. The old value/label still exists. For example: In contBinom, change Value 1 to 10, Undo: the 10 are still in the data. Change column, and come back to contBinom: you see still the label 10/10.

JorisGoosen replied:
aaah fuck yes

boutinb commented on CommonData/column.cpp:
```diff
@@ -1811,20 +1859,54 @@ Json::Value Column::serialize() const
+						-1
+			);
+}
+
```
Should the deserialize labels not be done inside a batch?

JorisGoosen replied:
Good one

boutinb commented on CommonData/databaseinterface.cpp:
```diff
@@ -21,14 +21,17 @@ void DatabaseInterface::upgradeDBFromVersion(Version originalVersion)
-			"ALTER TABLE Columns  ADD 	COLUMN emptyValuesJson TEXT;"		"\n"
-			"ALTER TABLE Columns  ADD 	COLUMN forceSourceColType INT NULL;""\n"
-			"ALTER TABLE Columns  DROP 	COLUMN isComputed;"					"\n"); // was removed in 0.18.3
+			R"C++20IsGreat(
```
It's c++11 feature!

JorisGoosen replied:
even more amazing ^^

boutinb reviewed (changes requested):
Still one bug: undo value change in Label Editor does not work. Cf below

JorisGoosen committed:
make sure label undo also undoes value changes in the data

JorisGoosen committed:
better stringhugger
